File: android/app/src/main/java/market/foodhome/app/telemetry/TelemetryEvent.kt
package market.foodhome.app.telemetry

enum class TelemetryEventName(val wireValue: String) {
    ShellLaunch("shell.launch"),
    WebViewLoadFailed("webview.load.failed"),
    WebViewRendererTerminated("webview.renderer.terminated"),
    BridgeRequestFailed("bridge.request.failed"),
    BridgeMethodUnsupported("bridge.method.unsupported"),
    BridgeVersionIncompatible("bridge.version.incompatible"),
    DeepLinkOpenFailed("deeplink.open.failed"),
    PushOpenFailed("push.open.failed"),
    PaymentReturnFailed("payment.return.failed"),
    MobileConfigFailed("mobile.config.failed"),
}

enum class TelemetryAttributeKey(val wireValue: String) {
    Platform("platform"),
    AppVersion("appVersion"),
    BridgeVersion("bridgeVersion"),
    WebViewVersion("webViewVersion"),
    RouteTemplate("routeTemplate"),
    ErrorCode("errorCode"),
    CorrelationId("correlationId"),
    NetworkClass("networkClass"),
    DurationBucket("durationBucket"),
}

data class TelemetryEvent internal constructor(
    val name: TelemetryEventName,
    val attributes: Map<TelemetryAttributeKey, String>,
)

File: ios/FoodHomeApp/Telemetry/TelemetryEvent.swift
import Foundation

enum TelemetryEventName: String, CaseIterable, Sendable {
    case shellLaunch = "shell.launch"
    case webViewLoadFailed = "webview.load.failed"
    case webViewRendererTerminated = "webview.renderer.terminated"
    case bridgeRequestFailed = "bridge.request.failed"
    case bridgeMethodUnsupported = "bridge.method.unsupported"
    case bridgeVersionIncompatible = "bridge.version.incompatible"
    case deepLinkOpenFailed = "deeplink.open.failed"
    case pushOpenFailed = "push.open.failed"
    case paymentReturnFailed = "payment.return.failed"
    case mobileConfigFailed = "mobile.config.failed"
}

enum TelemetryAttributeKey: String, CaseIterable, Sendable {
    case platform
    case appVersion
    case bridgeVersion
    case webViewVersion
    case routeTemplate
    case errorCode
    case correlationId
    case networkClass
    case durationBucket
}

struct TelemetryEvent: Equatable, Sendable, CustomStringConvertible {
    let name: TelemetryEventName
    let attributes: [TelemetryAttributeKey: String]

    var description: String {
        let values = attributes
            .map { "\($0.key.rawValue)=\($0.value)" }
            .sorted()
            .joined(separator: ",")
        return "\(name.rawValue){\(values)}"
    }
}
